{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e92a6f77-f5d3-4ee5-ae83-f5a21ba97218",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "#cat Homo_sapiens.GRCh38.114.gtf | awk '$3==\"gene\"' | tr ';' '\\t' | tr ' ' '\\t' | sed 's/\"//g' | cut -f1,4,5,7,10 | awk '{print $1\"\\t\"$2\"\\t\"$3\"\\t\"$5\"\\t0\\t\"$4}' > Homo_sapiens.GRCh38.114.bed\n",
    "\n",
    "# Import BED file from GTF file\n",
    "df = pd.read_csv(\"Homo_sapiens.GRCh38.114.bed\", sep=\"\\t\", header=None,\n",
    "                 names=[\"chr\", \"start\", \"end\", \"gene_id\", \"score\", \"strand\"])\n",
    "\n",
    "# Create promotor region +- 500 bp from TSS\n",
    "def get_promoter(row):\n",
    "    if row[\"strand\"] == \"+\":\n",
    "        tss = row[\"start\"]\n",
    "    else:\n",
    "        tss = row[\"end\"]\n",
    "    start = max(tss - 500, 0)  # unikamy ujemnych wartości\n",
    "    end = tss + 500\n",
    "    return pd.Series([row[\"chr\"], start, end, row[\"gene_id\"], row[\"strand\"]])\n",
    "\n",
    "promoters = df.apply(get_promoter, axis=1)\n",
    "promoters.columns = [\"chr\", \"prom_start\", \"prom_end\", \"gene_id\", \"strand\"]\n",
    "\n",
    "# Save BED file\n",
    "promoters[[\"chr\", \"prom_start\", \"prom_end\", \"gene_id\"]].to_csv(\"promoters_500bp.bed\",\n",
    "                                                               sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "88060c52-6f0c-4c41-8039-2c5bae72ce9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 1. Import cellid from Seurat stst file\n",
    "whitelist = pd.read_csv(\"k562.dis.stat.txt\", sep=\"\\t\", usecols=[\"cellid\"])\n",
    "cellids = set(whitelist[\"cellid\"])\n",
    "\n",
    "# 2. Import ATAC BED file made by bash command: cat k562.allele.flt.M.fragment.unsorted.tsv | sed 's/chr//' | cut -f1-4 > k562.allele.flt.M.fragment.unsorted.bed\n",
    "\n",
    "atac_cols = [\"chr\", \"start\", \"end\", \"barcode\", \"count\"]\n",
    "atac = pd.read_csv(\"k562.allele.flt.M.fragment.unsorted.bed\", sep=\"\\t\", names=atac_cols, header=None, dtype={\"chr\": str} )\n",
    "\n",
    "# 3. Filter ATAC by cellid\n",
    "atac_filtered = atac[atac[\"barcode\"].isin(cellids)]\n",
    "\n",
    "# 4. Save filtered file\n",
    "atac_filtered.to_csv(\"k562.atac.filtered.bed\", sep=\"\\t\", index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a1fa238-0167-4b7b-b4a8-ec43e0c01b52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make TSV file showiong localisation od ATC signal on chromosomes and in gene promotores\n",
    "\n",
    "#bedtools intersect -a k562.atac.filtered.bed -b promoters_500bp.bed -wa -wb > atac_mapped_to_genes.tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e94ad8a9-44df-4982-9fea-ec4f14c59955",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Read the ATAC localisation file\n",
    "cols = [\"chr_a\", \"start_a\", \"end_a\", \"barcode\", \"count\",\n",
    "        \"chr_b\", \"prom_start\", \"prom_end\", \"gene_id\"]\n",
    "mapped = pd.read_csv(\"atac_mapped_to_genes.tsv\", sep=\"\\t\", names=cols, header=None, dtype={\"chr_a\": str, \"chr_b\": str})\n",
    "\n",
    "# Count number of ATAC peaks for every gene in every cell\n",
    "atac_counts = mapped.groupby([\"barcode\", \"gene_id\"]).size().unstack(fill_value=0)\n",
    "\n",
    "# SAve matrix cells x genes\n",
    "atac_counts.to_csv(\"atac_counts_matrix.tsv\", sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d4e8b73-a512-4714-8619-5a42c8b3fbe7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a92df509-88f7-416b-a83f-c200d865bdfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NORMALIZACJA!?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "414c5e18-431a-4f80-9b8b-b66ecd34fa02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now import RNAseq expression and combine RNA with ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "263f7c32-d3f7-45a0-9f52-970ec508fe92",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_expr_norm = pd.read_csv(\"k562.allele.flt.M_df_expr_norm.tsv\", sep=\"\\t\", index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bd1d47a0-de59-4ed5-be05-25e675925462",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add Phase info to ATAC data\n",
    "atac_with_phase = atac_counts.merge(df_expr_norm[\"Phase\"], left_index=True, right_index=True)\n",
    "\n",
    "# Agregate ATAC by Phase (calculate mean ATAC signal)\n",
    "atac_by_phase = atac_with_phase.groupby(\"Phase\").mean()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0eeca859-621b-4575-8aa3-39ee4984a07c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001460</th>\n",
       "      <th>ENSG00000001561</th>\n",
       "      <th>ENSG00000002330</th>\n",
       "      <th>ENSG00000002726</th>\n",
       "      <th>ENSG00000002745</th>\n",
       "      <th>ENSG00000002822</th>\n",
       "      <th>ENSG00000002834</th>\n",
       "      <th>ENSG00000002919</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000310504</th>\n",
       "      <th>ENSG00000310507</th>\n",
       "      <th>ENSG00000310511</th>\n",
       "      <th>ENSG00000310515</th>\n",
       "      <th>ENSG00000310516</th>\n",
       "      <th>ENSG00000310525</th>\n",
       "      <th>ENSG00000310529</th>\n",
       "      <th>ENSG00000310564</th>\n",
       "      <th>ENSG00000310566</th>\n",
       "      <th>ENSG00000310567</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Phase</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G1</th>\n",
       "      <td>0.001220</td>\n",
       "      <td>0.000697</td>\n",
       "      <td>0.000871</td>\n",
       "      <td>0.000523</td>\n",
       "      <td>0.000871</td>\n",
       "      <td>0.001742</td>\n",
       "      <td>0.000697</td>\n",
       "      <td>0.003310</td>\n",
       "      <td>0.004704</td>\n",
       "      <td>0.002091</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008014</td>\n",
       "      <td>0.000697</td>\n",
       "      <td>0.002265</td>\n",
       "      <td>0.000348</td>\n",
       "      <td>0.000348</td>\n",
       "      <td>0.001916</td>\n",
       "      <td>0.000174</td>\n",
       "      <td>0.002265</td>\n",
       "      <td>0.012718</td>\n",
       "      <td>0.000697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G2M</th>\n",
       "      <td>0.000790</td>\n",
       "      <td>0.000474</td>\n",
       "      <td>0.001422</td>\n",
       "      <td>0.000632</td>\n",
       "      <td>0.000316</td>\n",
       "      <td>0.000790</td>\n",
       "      <td>0.000316</td>\n",
       "      <td>0.005216</td>\n",
       "      <td>0.003003</td>\n",
       "      <td>0.002687</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010590</td>\n",
       "      <td>0.001581</td>\n",
       "      <td>0.002529</td>\n",
       "      <td>0.001739</td>\n",
       "      <td>0.000316</td>\n",
       "      <td>0.002845</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.002055</td>\n",
       "      <td>0.011380</td>\n",
       "      <td>0.001106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.001247</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.000748</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.000624</td>\n",
       "      <td>0.001871</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>0.003741</td>\n",
       "      <td>0.002120</td>\n",
       "      <td>0.002993</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010849</td>\n",
       "      <td>0.000499</td>\n",
       "      <td>0.001372</td>\n",
       "      <td>0.000998</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.002494</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.002120</td>\n",
       "      <td>0.013717</td>\n",
       "      <td>0.001122</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 22140 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ENSG00000000460  ENSG00000000971  ENSG00000001460  ENSG00000001561  \\\n",
       "Phase                                                                       \n",
       "G1            0.001220         0.000697         0.000871         0.000523   \n",
       "G2M           0.000790         0.000474         0.001422         0.000632   \n",
       "S             0.001247         0.000249         0.000748         0.000125   \n",
       "\n",
       "       ENSG00000002330  ENSG00000002726  ENSG00000002745  ENSG00000002822  \\\n",
       "Phase                                                                       \n",
       "G1            0.000871         0.001742         0.000697         0.003310   \n",
       "G2M           0.000316         0.000790         0.000316         0.005216   \n",
       "S             0.000624         0.001871         0.000125         0.003741   \n",
       "\n",
       "       ENSG00000002834  ENSG00000002919  ...  ENSG00000310504  \\\n",
       "Phase                                    ...                    \n",
       "G1            0.004704         0.002091  ...         0.008014   \n",
       "G2M           0.003003         0.002687  ...         0.010590   \n",
       "S             0.002120         0.002993  ...         0.010849   \n",
       "\n",
       "       ENSG00000310507  ENSG00000310511  ENSG00000310515  ENSG00000310516  \\\n",
       "Phase                                                                       \n",
       "G1            0.000697         0.002265         0.000348         0.000348   \n",
       "G2M           0.001581         0.002529         0.001739         0.000316   \n",
       "S             0.000499         0.001372         0.000998         0.000249   \n",
       "\n",
       "       ENSG00000310525  ENSG00000310529  ENSG00000310564  ENSG00000310566  \\\n",
       "Phase                                                                       \n",
       "G1            0.001916         0.000174         0.002265         0.012718   \n",
       "G2M           0.002845         0.000158         0.002055         0.011380   \n",
       "S             0.002494         0.000249         0.002120         0.013717   \n",
       "\n",
       "       ENSG00000310567  \n",
       "Phase                   \n",
       "G1            0.000697  \n",
       "G2M           0.001106  \n",
       "S             0.001122  \n",
       "\n",
       "[3 rows x 22140 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_by_phase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1b40118a-b3fa-45db-b45b-c59490a9d94f",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_genes = avg_by_phase.columns.intersection(atac_by_phase.columns)\n",
    "\n",
    "rna_filtered = avg_by_phase.loc[:, common_genes]\n",
    "atac_filtered = atac_by_phase.loc[:, common_genes]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7078cd6c-c96f-494a-ab3c-93c14c187566",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make tables\n",
    "rna_long = rna_filtered.reset_index().melt(id_vars=\"Phase\", var_name=\"gene\", value_name=\"RNA_expr\")\n",
    "atac_long = atac_filtered.reset_index().melt(id_vars=\"Phase\", var_name=\"gene\", value_name=\"ATAC_access\")\n",
    "\n",
    "# Merge RNA and ATAC in one table\n",
    "merged = pd.merge(rna_long, atac_long, on=[\"Phase\", \"gene\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ba5b3dff-1d25-42df-b88c-c5069ad30817",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged.to_csv(\"k562.allele.flt.M_atacrnamerged.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "5edf215a-e196-48b2-b912-21b215abfff3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phase</th>\n",
       "      <th>gene</th>\n",
       "      <th>RNA_expr</th>\n",
       "      <th>ATAC_access</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G1</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>0.545158</td>\n",
       "      <td>0.001220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G2M</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>0.819479</td>\n",
       "      <td>0.000790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>0.949821</td>\n",
       "      <td>0.001247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G1</td>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>0.322822</td>\n",
       "      <td>0.000697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G2M</td>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>0.227182</td>\n",
       "      <td>0.000474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15841</th>\n",
       "      <td>G2M</td>\n",
       "      <td>ENSG00000288067</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15842</th>\n",
       "      <td>S</td>\n",
       "      <td>ENSG00000288067</td>\n",
       "      <td>0.008637</td>\n",
       "      <td>0.001122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15843</th>\n",
       "      <td>G1</td>\n",
       "      <td>ENSG00000288156</td>\n",
       "      <td>0.067966</td>\n",
       "      <td>0.000174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15844</th>\n",
       "      <td>G2M</td>\n",
       "      <td>ENSG00000288156</td>\n",
       "      <td>0.021988</td>\n",
       "      <td>0.000474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15845</th>\n",
       "      <td>S</td>\n",
       "      <td>ENSG00000288156</td>\n",
       "      <td>0.025090</td>\n",
       "      <td>0.000249</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15846 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Phase             gene  RNA_expr  ATAC_access\n",
       "0        G1  ENSG00000000460  0.545158     0.001220\n",
       "1       G2M  ENSG00000000460  0.819479     0.000790\n",
       "2         S  ENSG00000000460  0.949821     0.001247\n",
       "3        G1  ENSG00000000971  0.322822     0.000697\n",
       "4       G2M  ENSG00000000971  0.227182     0.000474\n",
       "...     ...              ...       ...          ...\n",
       "15841   G2M  ENSG00000288067  0.000000     0.001106\n",
       "15842     S  ENSG00000288067  0.008637     0.001122\n",
       "15843    G1  ENSG00000288156  0.067966     0.000174\n",
       "15844   G2M  ENSG00000288156  0.021988     0.000474\n",
       "15845     S  ENSG00000288156  0.025090     0.000249\n",
       "\n",
       "[15846 rows x 4 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
